Minimal work to make System.Diagnostics.Activity usable by the dotnet CLI application and codebase

baronfel opened the pull request:
I'm breaking out the monolithic set of changes from https://github.com/dotnet/sdk/pull/49409 into a more digestible series of changesets.

This first one adds the ability to _use_ System.Diagnostic.Activity at all in our codebase - creating an ActivitySource for us to use, and hooking the existing telemetry publishing up to it in a zero-cost way if no Activity Listeners are configured (which is the case today).

This is part of https://github.com/dotnet/sdk/issues/49668.

baronfel committed:
Minimal work to make System.Diagnostics.Activity usable by the dotnet CLI application and codebase

MiYanni reviewed (approved):
Looks good, but see the comments from @KalleOlaviNiemitalo

baronfel committed:
address dictionary-creation comments

don't return frozen dictionaries on modifiable call-sites. combine code for combining dictionaries
